src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -106,12 +110,31 @@ public int Count
             }
         }
 
-        public Enumerator GetEnumerator()
+        public T this[int index]
```

Comment:
`ref T` returning indexer will not work well with array covariance. Test of the ArraySegment is compatible with array covariance, so it will look odd.

```
String[] arr = new String[] { ... };
ArraySegment<object> segment = new ArraySegment(arr);
Object elem = segment[0]; // Throws exception if the indexer returns `ref T`
```